src/Common/ZooKeeper/ZooKeeper.cpp

```diff
@@ -55,6 +62,63 @@ static void check(Coordination::Error code, const std::string & path)
         throw KeeperException::fromPath(code, path);
 }
 
+ZooKeeperAvailabilityZoneMap& ZooKeeperAvailabilityZoneMap::instance()
+{
+    static ZooKeeperAvailabilityZoneMap map;
+    return map;
+}
+
+std::string ZooKeeperAvailabilityZoneMap::get(const std::string &host)
+{
+    std::lock_guard lock(mutex);
+    auto it = az_by_host.find(host);
+    if (it == az_by_host.end())
+        throw DB::Exception(DB::ErrorCodes::BAD_ARGUMENTS,"No availability zone info for host {}", host);
+    return it->second;
+}
+
+void ZooKeeperAvailabilityZoneMap::update(const std::string &host, const std::string &availability_zone)
+{
+    std::lock_guard lock(mutex);
+    az_by_host[host] = availability_zone;
+}
+
+std::vector<ShuffleHost> ZooKeeperAvailabilityZoneMap::shuffleHosts(const std::string & local_az, const std::vector<std::string> & hosts)
+{
+    std::vector<ShuffleHost> shuffled_hosts;
+    std::lock_guard lock(mutex);
+    for (size_t i = 0; i < hosts.size(); i++)
+    {
+        if (az_by_host.find(hosts[i]) == az_by_host.end())
+            az_by_host.emplace(hosts[i], AZ_UNKNWON);
+        ShuffleHost shuffle_host;
+        shuffle_host.host = hosts[i];
+        shuffle_host.original_index = static_cast<UInt8>(i);
+        if (local_az == az_by_host[hosts[i]])
+            shuffle_host.priority = Priority{0};
+        else if (az_by_host[hosts[i]] == AZ_UNKNWON)
+            shuffle_host.priority = Priority{1};
+        else
+            shuffle_host.priority = Priority{2};
+        shuffle_host.randomize();
+        shuffled_hosts.emplace_back(shuffle_host);
+    }
+    ::sort(shuffled_hosts.begin(), shuffled_hosts.end(), ShuffleHost::compare);
+    return shuffled_hosts;
+}
+
+bool ZooKeeperAvailabilityZoneMap::needTryOtherHost(const std::string &local_az, const std::set<std::string> &attempted_host)
```

Comment:
NOTE: make sure to document the caveat. if there're multiple keeper hosts, only focus on local az potentially could overload the remaining keeper host(s).

Reply:
Done.